### PR TITLE
Standing Orders submission and Reserved Accounts Compatibility 

### PR DESCRIPTION
**Pull request by josojo**
Takes care of #91 and #92 

Test plan:

unit tests only

### CODE REVIEW

File: contracts/SnappAuction.sol [hunk 1]
```diff
@@ -8,6 +8,16 @@ contract SnappAuction is SnappBase {
     uint16 public constant AUCTION_BATCH_SIZE = 1000;
     uint8 public constant AUCTION_RESERVED_ACCOUNTS = 50;
     uint8 public constant AUCTION_RESERVED_ACCOUNT_BATCH_SIZE = 10;
+    
+    struct StandingOrders {
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]:
```suggestion
    struct StandingOrder {
```

File: contracts/SnappAuction.sol [hunk 2]
```diff
@@ -8,6 +8,16 @@ contract SnappAuction is SnappBase {
...
+        bytes32 orderHash;
+        uint validityFrom;
+        uint validityTo;
+    }
+
+    // mapping from accountId to nounce to StandingOrders
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]:
```suggestion
    // mapping from accountId to nonce to StandingOrder
```

File: contracts/SnappAuction.sol (hunk 1 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: Since the hash contains several orders, I like the plural

File: contracts/SnappAuction.sol [hunk 3]
```diff
@@ -56,9 +75,61 @@ contract SnappAuction is SnappBase {
...
+
+    function getStandingOrderValidityTo(uint16 userId, uint128 nonce) public view returns (uint) {
+        return reservedAccountOrders[userId][nonce].validityTo;
+    }
+
     /**
      * Auction Functionality
      */
+    function placeStandingSellOrder(
+        uint8[] memory buyToken,
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: Why do we need memory here?

File: contracts/SnappAuction.sol [hunk 4]
```diff
@@ -8,6 +8,16 @@ contract SnappAuction is SnappBase {
     uint16 public constant AUCTION_BATCH_SIZE = 1000;
     uint8 public constant AUCTION_RESERVED_ACCOUNTS = 50;
     uint8 public constant AUCTION_RESERVED_ACCOUNT_BATCH_SIZE = 10;
+    
+    struct StandingOrders {
+        bytes32 orderHash;
+        uint validityFrom;
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: I would prefer the terms `validFrom` and `validTo`

File: contracts/SnappAuction.sol (hunk 3 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: its a memory, because we don't wanna store it, such that the contract can access it. And the newest solidity guidelines make me mention it explicitly

File: contracts/SnappAuction.sol [hunk 5]
```diff
@@ -56,9 +75,61 @@ contract SnappAuction is SnappBase {
...
+
     /**
      * Auction Functionality
      */
+    function placeStandingSellOrder(
+        uint8[] memory buyToken,
+        uint8[] memory sellToken,
+        uint128[] memory buyAmount,
+        uint128[] memory sellAmount
+    ) public onlyRegistered() {
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:110]: We might need to be stronger than only registered here... Maybe `onlySpecialAccount`.

File: contracts/SnappAuction.sol [hunk 6]
```diff
@@ -56,9 +75,61 @@ contract SnappAuction is SnappBase {
...
+        
+        require(auctionIndex < MAX_UINT, "Standing order collection has not yet started");
+
+        // Update Auction Hash based on request
+        uint16 accountId = publicKeyToAccountMap(msg.sender);
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:113]: `publicKeyToAccountMap` does not exist anymore. This must be an old instance. It is now called `registeredAccounts` and accessed in a very difference way.

Comment (bh2smith) [reply contracts/SnappAuction.sol]: `registeredAccounts`

File: contracts/SnappAuction.sol [hunk 7]
```diff
@@ -56,9 +75,61 @@ contract SnappAuction is SnappBase {
...
+        require(accountId <= AUCTION_RESERVED_ACCOUNTS, "Accout is not a rented account");
+
+        bytes32 orderHash;
+        uint nrOrder = buyToken.length;
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: suggest using a more common name here `numOrders`

File: contracts/SnappAuction.sol (hunk 3 above)
Comment (bh2smith) [inline contracts/SnappAuction.sol]: all of these input parameters should be plural (e.g. `buyTokens`).

File: contracts/SnappAuction.sol (hunk 1 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Oh yes. right

File: contracts/SnappAuction.sol (hunk 5 above)
Comment (josojo) [reply contracts/SnappAuction.sol:110]: `require(accountId <= AUCTION_RESERVED_ACCOUNTS, "Accout is not a reserved account");`
 takes care of this

File: contracts/SnappAuction.sol (hunk 6 above)
Comment (josojo) [reply contracts/SnappAuction.sol:113]: it's even used in the placeSellOrder function

File: contracts/SnappAuction.sol (hunk 1 above)
Comment (fleupold) [reply contracts/SnappAuction.sol]: How about `StandingOrderBatch` ?

File: contracts/SnappAuction.sol [hunk 8]
```diff
@@ -8,6 +8,16 @@ contract SnappAuction is SnappBase {
     uint16 public constant AUCTION_BATCH_SIZE = 1000;
     uint8 public constant AUCTION_RESERVED_ACCOUNTS = 50;
     uint8 public constant AUCTION_RESERVED_ACCOUNT_BATCH_SIZE = 10;
+    
+    struct StandingOrders {
+        bytes32 orderHash;
+        uint validFrom;
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Can we somehow make clear that this is an auctionIndex? Also is this an inclusive or exclusive range?

File: contracts/SnappAuction.sol [hunk 9]
```diff
@@ -8,6 +8,16 @@ contract SnappAuction is SnappBase {
...
+        uint validTo;
+    }
+
+    //mapping from accountId to nonce to StandingOrders
+    mapping (uint16 => mapping(uint128 => StandingOrders)) public reservedAccountOrders;
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Would it be better to model this as a single datatype (mapping from accountId to tuple of currentNonce)?

`mapping(uint16 => (uint128, mapping(uint128 => StandingOrders))`

This would save storing the `uint16` twice.

Also curious, how you decided to use uint128? Should we consistent with the datatypes we already have and either use `uint32` ore `uint256`?

+ reserved account can be `uint8`

File: contracts/SnappAuction.sol (hunk 5 above)
Comment (fleupold) [reply contracts/SnappAuction.sol:110]: Could be nice to make this a modifier to reuse?

File: contracts/SnappAuction.sol [hunk 10]
```diff
@@ -22,6 +32,15 @@ contract SnappAuction is SnappBase {
         uint128 sellAmount
     );
 
+    event StandingSellOrder(
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: I think it would be better to emit all orders of the micro batch in one event. This is cheaper from a gas perspective and since no other function would use this event individually (as in it would always a batch of 10), combining data into one event would also make database writes more atomic.

File: contracts/SnappAuction.sol [hunk 11]
```diff
@@ -56,9 +75,66 @@ contract SnappAuction is SnappBase {
...
+            block.timestamp > (auctions[auctionIndex].creationTimestamp + 3 minutes)
+        ) {
+            createNewAuctionBatch();
+        }
+
+        for (uint i = 0; i < numOrders; i++) {
+            // Must have 0 < tokenId < MAX_TOKENS anyway, so may as well ensure registered.
+            require(buyTokens[i] < numTokens, "Buy token is not registered");
+            require(sellTokens[i] < numTokens, "Sell token is not registered");
+            orderHash = sha256(
```
Comment (fleupold) [inline contracts/SnappAuction.sol:128]: No need to be rolling here. This can just be `sha(all_order_concatenated)`

File: contracts/SnappAuction.sol [hunk 12]
```diff
@@ -56,9 +75,66 @@ contract SnappAuction is SnappBase {
...
+        if (
+            auctionIndex == MAX_UINT ||
+            block.timestamp > (auctions[auctionIndex].creationTimestamp + 3 minutes)
+        ) {
+            createNewAuctionBatch();
+        }
+
+        for (uint i = 0; i < numOrders; i++) {
+            // Must have 0 < tokenId < MAX_TOKENS anyway, so may as well ensure registered.
+            require(buyTokens[i] < numTokens, "Buy token is not registered");
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: nit: maybe we want to replace this with a checkOrderValid function that works on the encoded order and can be reused for standing orders and individual orders?

File: contracts/SnappAuction.sol [hunk 13]
```diff
@@ -153,4 +219,18 @@ contract SnappAuction is SnappBase {
     function maxUnreservedOrderCount() internal pure returns (uint16) {
         return AUCTION_BATCH_SIZE - (AUCTION_RESERVED_ACCOUNTS * AUCTION_RESERVED_ACCOUNT_BATCH_SIZE);
     }
+
+    function createNewAuctionBatch() internal {
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: nit: maybe name this createNewPendingBatch() to avoid introducing another name?

File: contracts/SnappAuction.sol [hunk 14]
```diff
@@ -76,17 +152,7 @@ contract SnappAuction is SnappBase {
             auctions[auctionIndex].size == maxUnreservedOrderCount() ||
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: I'm wondering if we want to drop this condition, since otherwise, we could prevent people from cancelling order by filling the batch with non-sense unreserved orders (of course only until we implement a fee model)?

File: contracts/SnappAuction.sol [hunk 15]
```diff
@@ -56,9 +75,66 @@ contract SnappAuction is SnappBase {
...
+            require(sellTokens[i] < numTokens, "Sell token is not registered");
+            orderHash = sha256(
+                abi.encodePacked(
+                    orderHash,
+                    encodeOrder(accountId, buyTokens[i], sellTokens[i], buyAmounts[i], sellAmounts[i])
+                )
+            );
+            emit StandingSellOrder(auctionIndex, accountId, buyTokens[i], sellTokens[i], buyAmounts[i], sellAmounts[i]);
+        }
+        uint128 currentNonce = standingOrderNonce[accountId];
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: This code doesn't seem to be covered by any unit test.

File: contracts/SnappAuction.sol [hunk 16]
```diff
@@ -56,9 +75,66 @@ contract SnappAuction is SnappBase {
...
+        if (auctionIndex > 0) {
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Would it be better/simpler to override the currentNonce if the `validFrom == auctionIndex` (that is multiple updates per block)?

File: contracts/SnappAuction.sol [hunk 17]
```diff
@@ -56,9 +75,66 @@ contract SnappAuction is SnappBase {
...
+            reservedAccountOrders[accountId][currentNonce].validTo = auctionIndex - 1;
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: in case `auctionIndex == 1` it will be hard to differentiate this entry from an "uncancelled" entry (where validTo is also 0)

File: contracts/SnappAuction.sol (hunk 3 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Right I noticed when I tried to test my mult-order thing.

**Review by fleupold — COMMENTED** [review]
Really good progress. I'm surprised the two tasks fit in one PR, but it's not yet that large.

A few high-level discussion points:

- Modeling of the reserved account batch mapping
- Should we compute an iterative rolling hash or a single hash for reserved batches (`sha(sha(...sha(o1), ..., o1)` vs `sha(o1, ..., o10)`)?
- Should there be a single event per batch each update instead of up to 10?
- The nonce/datatype update logic needs testing and some more clarifications

File: test/snapp_auction.js [hunk 18]
```diff
@@ -128,6 +128,67 @@ contract("SnappAuction", async (accounts) => {
...
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+
+      await instance.placeStandingSellOrder([0,0], [0,1], [3,1], [3,1], { from: user_1 })
+
+      const userId = await instance.publicKeyToAccountMap.call(user_1)
+      const nonce = await instance.standingOrderNonce.call(userId)
+      assert.equal(nonce, 1)
+    })
+  })
```
Comment (bh2smith) [inline test/snapp_auction.js:244]: Should probably include a test with more than 10 orders submitted.

File: contracts/SnappAuction.sol [hunk 19]
```diff
@@ -56,9 +75,76 @@ contract SnappAuction is SnappBase {
...
+        uint128[] memory sellAmount
+    ) public onlyRegistered() {
+        
+        // Update Auction Hash based on request
+        uint16 accountId = publicKeyToAccountMap(msg.sender);
+        require(accountId <= AUCTION_RESERVED_ACCOUNTS, "Accout is not a rented account");
+
+        bytes32 orderHash;
+        uint nrOrder = buyToken.length;
+        require(nrOrder <= AUCTION_RESERVED_ACCOUNT_BATCH_SIZE, "Too many orders");
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: This an insufficient number of requirements. 

```js
        uint numOrders = buyTokens.length;
        require(numOrders == sellTokens.length, "numOrders != sellTokens length");
        require(numOrders == buyAmounts.length, "numOrders != buyAmounts length");
        require(numOrders == sellAmounts.length, "numOrders != sellAmounts length");
```

File: contracts/SnappAuction.sol [hunk 20]
```diff
@@ -56,9 +75,76 @@ contract SnappAuction is SnappBase {
...
+        
+        if (
+            auctionIndex == MAX_UINT ||
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:121]: If solidity is a "short circuiting" I would think that we could be more efficient if the auctionIndex check occurred second. I really don't like this checking the contract initialization case every single time.

File: contracts/SnappAuction.sol [hunk 21]
```diff
@@ -56,9 +75,76 @@ contract SnappAuction is SnappBase {
         return auctions[slot].appliedAccountStateIndex != 0;
     }
 
+    function getStandingOrderHash(uint16 userId, uint128 nonce) public view returns (bytes32) {
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: Do we need all these getters? Or is this just expected for the future?

File: contracts/SnappAuction.sol [hunk 22]
```diff
@@ -56,9 +75,76 @@ contract SnappAuction is SnappBase {
...
+            orderHash = sha256(
+                abi.encodePacked(
+                    orderHash,
+                    encodeOrder(accountId, buyToken[i], sellToken[i], buyAmount[i], sellAmount[i])
+                )
+            );
+            emit StandingSellOrder(auctionIndex, accountId, buyToken[i], sellToken[i], buyAmount[i], sellAmount[i]);
+        }
+        uint128 currentNonce = standingOrderNonce[accountId];
+        if (auctionIndex > 0) {
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: So, essentially always, this is what happens.

File: contracts/SnappAuction.sol [hunk 23]
```diff
@@ -56,9 +75,76 @@ contract SnappAuction is SnappBase {
...
+            reservedAccountOrders[accountId][currentNonce].validTo = auctionIndex - 1;
+        } else {
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: What is being deleted here?

File: contracts/SnappAuction.sol [hunk 24]
```diff
@@ -56,9 +75,76 @@ contract SnappAuction is SnappBase {
...
+            delete reservedAccountOrders[accountId][currentNonce];
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: What are we deleting?

File: contracts/SnappAuction.sol [hunk 25]
```diff
@@ -56,9 +75,76 @@ contract SnappAuction is SnappBase {
...
+            // Must have 0 < tokenId < MAX_TOKENS anyway, so may as well ensure registered.
+            require(buyToken[i] < numTokens, "Buy token is not registered");
+            require(sellToken[i] < numTokens, "Sell token is not registered");
+            orderHash = sha256(
+                abi.encodePacked(
+                    orderHash,
+                    encodeOrder(accountId, buyToken[i], sellToken[i], buyAmount[i], sellAmount[i])
+                )
+            );
+            emit StandingSellOrder(auctionIndex, accountId, buyToken[i], sellToken[i], buyAmount[i], sellAmount[i]);
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: I believe this even should include an index!

File: contracts/SnappAuction.sol (hunk 8 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: `auctionStart` and `auctionEnd`

File: contracts/SnappAuction.sol (hunk 12 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Agree!

File: contracts/SnappAuction.sol (hunk 17 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Good catch.

File: contracts/SnappAuction.sol (hunk 14 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: How would we ensure the batch isn't maxed out without it. People also shouldn't be able to cancel generic orders.

File: contracts/SnappAuction.sol (hunk 19 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: It will throw error automatically. I could write this down, but it would just be more code to digest

File: contracts/SnappAuction.sol (hunk 21 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: It is expected for the future and I want to add some more tests

File: contracts/SnappAuction.sol (hunk 9 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: the uint16 is not stored technically, it is just hashed.
I would take the solution, which is more readable, although yours is more gas efficient. But I am not sure, which one is more readable. 

Yeah, I will change to uint256. 

I used uint16 for reserved accounts, so that I do not have to do the transformation from uint16 to uint8 after calling this function:
>        uint16 accountId = publicKeyToAccountMap(msg.sender);

Storing as a unit8 would not make any gas difference, as this value gets hashed anyways and both values are below a wordsize

File: contracts/SnappAuction.sol (hunk 14 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: Yeah, not sure what to do here...

Probably, I would keep it in here, to avoid new logic. And the current logic (or similar) 
will have to be reimplemented anyways once we have the fee model.

File: contracts/SnappAuction.sol (hunk 19 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: Yeah, let's go without this in the current PR and then write a new PR changing our code to take orderblobs instead of several parameters

File: contracts/SnappAuction.sol (hunk 9 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: > mapping(uint16 => (uint128, mapping(uint128 => StandingOrders))

checked the readablity. Does not look too shabby. Lets go with it.

File: contracts/SnappAuction.sol (hunk 11 above)
Comment (josojo) [reply contracts/SnappAuction.sol:128]: is it?
I guess the validation proofs are simpler if we do the rolling hash. If we do not do the rolling hash, we would have to hand over the order numbers of each of the 50 batches inorder to reconstruct the hash. If we do the rolling hash, we can just hand over the data as one blob and keep on calculating the iterative hash each x bytes

File: contracts/SnappAuction.sol [hunk 26]
```diff
@@ -9,15 +9,19 @@ contract SnappAuction is SnappBase {
     uint8 public constant AUCTION_RESERVED_ACCOUNTS = 50;
     uint8 public constant AUCTION_RESERVED_ACCOUNT_BATCH_SIZE = 10;
     
-    struct StandingOrders {
+    struct StandingOrderBatch {
         bytes32 orderHash;
-        uint validityFrom;
-        uint validityTo;
+        uint validFromAuctionIndex;
+        uint validToAuctionIndex;
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Can we please comment here if they are inclusive or exclusive?

File: contracts/SnappAuction.sol [hunk 27]
```diff
@@ -32,13 +36,14 @@ contract SnappAuction is SnappBase {
         uint128 sellAmount
     );
 
-    event StandingSellOrder(
-        uint validityFrom, 
+    event StandingSellOrderBatch(
+        uint validFromAuctionIndex,
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Why is this needed? As far as the database is concerned it will probably store either nonce or validFromAuctionIndex (depending on whether there can be two entries with the same validFromAuctionIndex). But I don't think we need both. I'd suggest to use nonce, since it's unique by definition

File: contracts/SnappAuction.sol [hunk 28]
```diff
@@ -76,58 +81,66 @@ contract SnappAuction is SnappBase {
...
         // Update Auction Hash based on request
         uint16 accountId = publicKeyToAccountMap(msg.sender);
-        require(accountId <= AUCTION_RESERVED_ACCOUNTS, "Accout is not a rented account");
+        require(accountId <= AUCTION_RESERVED_ACCOUNTS, "Accout is not a reserved account");
 
         bytes32 orderHash;
-        uint nrOrder = buyToken.length;
-        require(nrOrder <= AUCTION_RESERVED_ACCOUNT_BATCH_SIZE, "Too many orders");
+        uint numOrders = buyTokens.length;
+        require(numOrders <= AUCTION_RESERVED_ACCOUNT_BATCH_SIZE, "AUCTION_RESERVED_ACCOUNT_BATCH_SIZE not considered");
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: nit: I liked the old wording which I think was more clear and didn't leak internal constant names. Maybe "Too many orders for reserved batch"?

File: contracts/SnappAuction.sol (hunk 11 above)
Comment (fleupold) [reply contracts/SnappAuction.sol:128]: First, I think we will end up calculating a large hash over all 1000 orders for the fraud proof anyways instead of rolling all of them.

But even if not, we can still submit one large blob, roll-hash the first 500 items and then bulk-hash the remaining 500 in 10 increments. I don't see how "order numbers" are required to reconstruct the hash?

One argument in favour of the current approach is that the code in the driver can be completely reused whereas otherwise we have to extend the hashing logic.

Yet, I think it's unnecessary to roll and would make the logic of standing order collection simpler.

File: contracts/SnappAuction.sol [hunk 29]
```diff
@@ -76,58 +81,66 @@ contract SnappAuction is SnappBase {
...
-            emit StandingSellOrder(auctionIndex, accountId, buyToken[i], sellToken[i], buyAmount[i], sellAmount[i]);
-        }
-        uint128 currentNonce = standingOrderNonce[accountId];
-        if (auctionIndex > 0) {
-            reservedAccountOrders[accountId][currentNonce].validityTo = auctionIndex - 1;
-        } else {
-            delete reservedAccountOrders[accountId][currentNonce];
+        }        
+        uint currentNonce = standingOrderPlacements[accountId].standingOrderNonce;
+        uint newNonce = currentNonce;
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: nit: not needed. Could just be currentNonce?

File: contracts/SnappAuction.sol [hunk 30]
```diff
@@ -76,58 +81,66 @@ contract SnappAuction is SnappBase {
...
+        
+        if (auctionIndex > standingOrderPlacements[accountId].reservedAccountOrders[currentNonce].validFromAuctionIndex) {
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: nit: I wonder if it would be more readable to temporarily store the StandingOrderData struct and current StandingSellOrderBatch instead of writing these long accessors multiple times?

**Review by fleupold — APPROVED** [review]
Looks good overall.

I don't really like that we are rolling the standing orders for no obvious reason and that standing order submission happens with multiple array which have to implicitly be aligned.

However, I'd be ok to change this in a separate PR to unblock the rest of the work, but would still consider it part of #92.

File: contracts/SnappAuction.sol (hunk 26 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Holy macaroni, the words here just got way longer what was wrong with `auctionStart` and `auctionEnd` or `auctionFrom`-`auctionTo`

@fleupold would it suffice to include the word inclusive in a comment immediately after the declaration?

For example, 

```js
   uint auctionStart;  // inclusive auction index of validity
   uint auctionEnd;    // also inclusive 
```

File: contracts/SnappAuction.sol (hunk 28 above)
Comment (bh2smith) [reply contracts/SnappAuction.sol]: I would also be in favour of "Too many orders for reserved batch"

Comment (bh2smith) [reply contracts/SnappAuction.sol]:
```suggestion
        require(numOrders <= AUCTION_RESERVED_ACCOUNT_BATCH_SIZE, "Too many orders for reserved batch");
```

File: contracts/SnappAuction.sol [hunk 31]
```diff
@@ -140,17 +209,37 @@ contract SnappAuction is SnappBase {
         uint128 buyAmount, 
         uint128 sellAmount
     ) 
-        internal pure returns (bytes32) 
+        internal view returns (bytes32)
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:218]: @josojo  can you please explain the change here? I see it is conflicting with master and wonder which to choose?

Comment (bh2smith) [reply contracts/SnappAuction.sol:218]: I guess the changes that have been made here are no longer "pure"

Comment (josojo) [reply contracts/SnappAuction.sol:218]: yeah, if you use variables from inherited contracts, you can no longer use pure.

File: contracts/SnappAuction.sol (hunk 26 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: I don't like auctionStart and auctionEnd to much, as its a statement about validity rather than a start time.

File: contracts/SnappAuction.sol (hunk 29 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: I thought it would increase readability.

**Review by bh2smith — APPROVED** [review]
Nice job!

File: contracts/SnappAuction.sol [hunk 32]
```diff
@@ -22,6 +36,15 @@ contract SnappAuction is SnappBase {
         uint128 sellAmount
     );
 
+    event StandingSellOrderBatch(
+        uint validFromAuctionIndex,
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Thinking about this again, do we then even need a nonce? If validFromAuctionIndex uniquely identifies the batch what's the point of nonce?

Comment (fleupold) [reply contracts/SnappAuction.sol]: I guess once we allow rollbacks validFromAuction could become smaller than a previous value. in that case, I believe we should emit nonce as part of the event and not the auction index.

Comment (josojo) [reply contracts/SnappAuction.sol]: The smart contract needs to know the previous ordersubmission, in order to set the validToAuctionIndex.

But of course, this data can be submitted as payload to the function. After reading your recommended book, I did not implement it, in order to keep the code simpler. Maybe I should, as this would save gas costs.

Yeah, I am not sure, how relevant the rollbacks are. I would suggest that traders are supposed to verify the state and do not submit orders if the state is illegal. If they do, then they are committed to those orders, even when auctions are reversed.

What do you guys think, should I implement that the previous nonce is handed over via payload and verified by checking that the validToAuctionIndex of this orderBatch is not set?
@fleupold @bh2smith

Comment (josojo) [reply contracts/SnappAuction.sol]: Hm, let me do a quick PR into this PR.
Having the code is easier to discuss.

Comment (josojo) [reply contracts/SnappAuction.sol]: I created the PR

Comment (bh2smith) [reply contracts/SnappAuction.sol]: I don't know what is happening here anymore.

**Comment by josojo** [comment]
I hope now everything is as discussed. Could someone review it one more time? 

One thing that was not explicitly mentioned is the removal of validToIndex variable. But I am getting behind Ben's opinion that is is not really needed.

File: contracts/SnappAuction.sol [hunk 33]
```diff
@@ -56,37 +78,92 @@ contract SnappAuction is SnappBase {
...
+
     /**
      * Auction Functionality
      */
+    function placeStandingSellOrder(
+        uint8[] memory buyTokens,
+        uint8[] memory sellTokens,
+        uint128[] memory buyAmounts,
+        uint128[] memory sellAmounts,
+        uint validTill
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: `validUntil`? Generally, can we try not to abbreviate variables unless absolutely necessary? It's slightly faster for you to type now, but harder for everyone who reads the code to understand.

File: contracts/SnappAuction.sol [hunk 34]
```diff
@@ -8,6 +8,19 @@ contract SnappAuction is SnappBase {
...
+    
+    struct StandingOrderBatch {
+        bytes32 orderHash;
+        uint validFromIndex; // validity is inclusive of the auction index
+        // validToIndex is indirectly given as validFromIndex of next Orderbatch
+    }
+
+    struct StandingOrderData {
+        mapping(uint => StandingOrderBatch) reservedAccountOrders;
+        uint currentCnt;
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: currentCount or counter please. Why not nonce or currentBatchIndex/currentBatchKey? I find count a bit ambiguous.

File: contracts/SnappAuction.sol [hunk 35]
```diff
@@ -56,37 +78,92 @@ contract SnappAuction is SnappBase {
         return auctions[slot].appliedAccountStateIndex != 0;
     }
 
+    function getStandingOrderHash(uint16 userId, uint128 current) public view returns (bytes32) {
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: can we name `current` a bit more descriptively. e.g. `batchIndex` since it could technically also be an old one.

File: contracts/SnappAuction.sol [hunk 36]
```diff
@@ -22,6 +35,15 @@ contract SnappAuction is SnappBase {
         uint128 sellAmount
     );
 
+    event StandingSellOrderBatch(
+        uint validFromIndex,
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: Since we rely on `currentCnt` in the public view functions, I'd argue it would be nicer to emit that rather than the `validFromIndex`. If needed `validFrom` can be queried by using `getStandingOrderValidFrom`

File: contracts/SnappAuction.sol [hunk 37]
```diff
@@ -56,37 +78,92 @@ contract SnappAuction is SnappBase {
...
+        for (uint i = 0; i < numOrders; i++) {
+            orderHash = sha256(
+                abi.encodePacked(
+                    orderHash,
+                    encodeOrder(accountId, buyTokens[i], sellTokens[i], buyAmounts[i], sellAmounts[i])
+                )
+            );
+        }        
+        uint currentCnt = standingOrderData[accountId].currentCnt;
+        StandingOrderBatch memory standingOrderBatch = standingOrderData[accountId].reservedAccountOrders[currentCnt];
```
Comment (fleupold) [inline contracts/SnappAuction.sol]: I guess we can wait until we have rollbacks to actually require that `auctionIndex >= standingOrderBatch.validFromIndex`. Wondering if it makes sense to capture that somehow now so that we don't forget?

**Review by fleupold — COMMENTED** [review]
Some comments around naming. Also, I think we should emit the pointer to the current standingOrderBatch in event rather than the valid from index. The latter can easily be looked up by the former but not vice versa.

File: contracts/SnappAuction.sol (hunk 37 above)
Comment (josojo) [reply contracts/SnappAuction.sol]: No, it is even needed for today. If people are overwriting their orders several times during one auction index, we also do not need to increase the counter etc. We only need to update the orderHash.
I think having it like it is coded is much more efficient.

File: contracts/SnappAuction.sol [hunk 38]
```diff
@@ -8,6 +8,19 @@ contract SnappAuction is SnappBase {
...
+        uint validFromIndex; // validity is inclusive of the auction index
+        // validToIndex is indirectly given as validFromIndex of next Orderbatch
+    }
+
+    struct StandingOrderData {
+        mapping(uint => StandingOrderBatch) reservedAccountOrders;
+        uint currentBatchIndex;
+    }
+
+    mapping (uint16 => StandingOrderData) public standingOrderData;
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: My only worry here is using `standingOrderData` and `StandingOrderData` for two different but similar things. I would propose either that 
- `StandingOrderData` become `StandingOrder` or
- `standingOrderData` become `standingOrders` or `standingOrderMap`

**Review by bh2smith — APPROVED** [review]
Seems alright, not quite sure about the wording yet... We have `standingOrderData`, `StandingOrderData` and `StandingOrderBatch` all meaning subtly different things.

File: contracts/SnappAuction.sol [hunk 39]
```diff
@@ -56,37 +78,90 @@ contract SnappAuction is SnappBase {
...
+
+        bytes32 orderHash;
+        uint numOrders = buyTokens.length;
+        require(numOrders <= AUCTION_RESERVED_ACCOUNT_BATCH_SIZE, "Too many orders for reserved batch");
+        
+        if (
+            auctionIndex == MAX_UINT ||
+            block.timestamp > (auctions[auctionIndex].creationTimestamp + 3 minutes)
+        ) {
+            createNewPendingBatch();
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:124]: Why are we doing this here? These batches are not affected by the code that follows. I see no reason to update the unreserved order batch.

Comment (josojo) [reply contracts/SnappAuction.sol:124]: answered in the issue
